dotCMS/src/main/java/com/dotcms/business/SystemTableFactoryImpl.java

```diff
@@ -0,0 +1,136 @@
+
+    private final SystemCache systemCache;
+
+    private final static String VALUE_404 = "DOT_SYSTEM_TABLE_404";
+
+    public SystemTableFactoryImpl () {
+
+        this.systemCache = CacheLocator.getSystemCache();
+    }
+    @Override
+    protected Optional<String> find(final String key) throws DotDataException {
+
+        Object value = null;
+        if(UtilMethods.isSet(key)) {
+
+            value = this.systemCache.get(key);
+            if(Objects.isNull(value)) {
+
+                final List<Map<String, Object>> result = new DotConnect()
+                        .setSQL(" SELECT * FROM system_table WHERE key = ? ")
+                        .addParam(key)
+                        .loadObjectResults();
+
+                value = result.isEmpty() ? null : result.get(0).get("value");
+
+                if(Objects.nonNull(value)) {
+                    this.systemCache.put(key, value);
+                } else {
+                    this.systemCache.put(key, VALUE_404);
+                    value = VALUE_404;
+                }
+            }
+        }
+
+        return  Objects.isNull(value) || VALUE_404.equals(value)?
+                Optional.empty():
+                Optional.ofNullable(toString(value));
+    }
+
+    @Override
+    protected Map<String, String> findAll() throws DotDataException {
+
+        final Map<String, String> records = new LinkedHashMap<>();
+
+        final List<Map<String, Object>> result = new DotConnect()
+                .setSQL(" SELECT * FROM system_table")
+                .loadObjectResults();
+
+        for (final Map<String, Object> recordMap : result) {
+
+            final Object key = recordMap.get("key");
+            final Object value = recordMap.get("value");
+            if (Objects.nonNull(key) && Objects.nonNull(value)) {
+                records.put(key.toString(), toString(value.toString()));
+                this.systemCache.put(key.toString(), value);
+            }
+        }
+
+        return records;
+    }
+
+    @Override
+    protected void saveOrUpdate(final String key, final String value) throws DotDataException {
+
+        if (Objects.nonNull(key) && Objects.nonNull(value)) {
+
+            final Optional<String> valueOpt = find(key);
+
+            new DotConnect()
+                    .setSQL(valueOpt.isPresent()?
+                            "UPDATE system_table SET value=? WHERE key=?":
+                            "INSERT INTO system_table (value, key) VALUES (?,?)")
+                    .addParam(value)
+                    .addParam(key)
+                    .loadResult();
+
+            this.systemCache.remove(key);
+        } else {
+
+            throw new DotDataException("The key and value should not be null");
+        }
+    }
+
+    @Override
+    protected void delete(final String key) throws DotDataException {
+
+        if (Objects.nonNull(key)) {
+
+            final Optional<String> valueOpt = find(key);
```

Comment:
I would just run the delete no matter what - rather than check for the existence of the key, which opens up room for race conditions.

Reply:
why throw an error if it does not exist?  who cares?